service/s3/s3crypto/crypto_mode.go

```diff
@@ -0,0 +1,138 @@
+package s3crypto
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+// CryptoMode is an abstraction layer that deals with encryption of
+// the contents and which key provider to use.
+type CryptoMode interface {
+	EncryptContents(io.Writer, io.Reader) error
+	DecryptContents(KeyProvider, io.ReadCloser) (io.ReadCloser, error)
+	GetKeyProvider() KeyProvider
+	CipherDataIface
+}
+
+// CipherDataIface is used for when populating the envelope details upon
+// encryption.
+type CipherDataIface interface {
+	GetCipherName() string
+	GetTagLen() string
+}
+
+// DecryptMode is meant to used only in reading objects from s3
+type DecryptMode interface {
+	DecryptContents(KeyProvider, io.ReadCloser) (io.ReadCloser, error)
+	GetKeyProvider() KeyProvider
+}
+
+func modeFactory(env *Envelope, cfg Config) (DecryptMode, error) {
+	kp, err := keyProviderFactory(env, cfg)
```

Comment:
Nit I'd rename `modeFactory` to `modeFromEnvelope`. Only because its more descriptive.